test/test_dir.rs

```diff
@@ -0,0 +1,49 @@
+extern crate nix;
+extern crate tempdir;
+
+use nix::dir::{Dir, Type};
+use std::fs::File;
+use self::tempdir::TempDir;
+
+fn open(path: &::std::path::Path) -> nix::Result<Dir> {
+    use nix::fcntl::OFlag;
+    let fd = nix::fcntl::open(
+        path,
+        OFlag::O_DIRECTORY | OFlag::O_RDONLY | OFlag::O_CLOEXEC,
+        nix::sys::stat::Mode::empty()
+    ).unwrap();
+    Dir::from_fd(fd)
+}
+
+#[test]
+fn read() {
+    let tmp = TempDir::new("nix-dir-read").unwrap();
+    File::create(&tmp.path().join("foo")).unwrap();
+    ::std::os::unix::fs::symlink("foo", tmp.path().join("bar")).unwrap();
+    let mut dir = open(tmp.path()).unwrap();
+    let mut entries: Vec<_> = dir.into_iter().map(|e| e.unwrap()).collect();
+    entries.sort_by(|a, b| a.file_name().cmp(b.file_name()));
+    let entry_names: Vec<_> = entries
+        .iter()
+        .map(|e| e.file_name().to_str().unwrap().to_owned())
+        .collect();
+    assert_eq!(&entry_names[..], &[".", "..", "bar", "foo"]);
+    assert!(&[Some(Type::Directory), None].contains(&entries[0].type_())); // .: dir
```

Comment:
It would be worthwhile adding a comment explaining why the check for None is necessary.

Reply:
Added.